File: 2015m./uzd2./sprendimai/uzd2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
using namespace std;

class Sheep {
public:
    int geneticSimilarity;
    string name;
    string dnr;

    Sheep(string name, string dnr) : name(name), dnr(dnr), geneticSimilarity(0) {}
};

vector<Sheep> readData(int& geneCount, Sheep& studiedSheep) {
    ifstream data("U2.txt");
    int sheepCount;
    data >> sheepCount >> geneCount;
    int studiedSheepIndex;
    data >> studiedSheepIndex;
    studiedSheepIndex--;

    vector<Sheep> sheep;

    for (int i = 0; i < sheepCount; ++i) {
        string name, dnr;
        data >> name >> dnr;
        if (i == studiedSheepIndex) {
            studiedSheep = Sheep(name, dnr);
        } else {
            sheep.push_back(Sheep(name, dnr));
        }
    }

    data.close();

    return sheep;
}

void outputData(const Sheep& studiedSheep, const vector<Sheep>& sheep) {
    ofstream output("U2rez.txt");
    output << studiedSheep.name << endl;
    for (const Sheep& s : sheep) {
        output << s.name << " " << s.geneticSimilarity << endl;
    }
    output.close();
}

vector<Sheep> sortSheep(vector<Sheep> sheep) {
    int size = sheep.size();
    for (int i = 0; i < size; ++i) {
        for (int j = i + 1; j < size; ++j) {
            if (sheep[i].geneticSimilarity < sheep[j].geneticSimilarity ||
                (sheep[i].geneticSimilarity == sheep[j].geneticSimilarity && sheep[i].name > sheep[j].name)) {
                swap(sheep[i], sheep[j]);
            }
        }
    }
    return sheep;
}

int getGeneticSimilarity(const string& dnr1, const string& dnr2, int geneCount) {
    int geneticSimilarity = 0;
    for (int i = 0; i < geneCount; ++i) {
        if (dnr1[i] == dnr2[i]) {
            geneticSimilarity++;
        }
    }
    return geneticSimilarity;
}

int main() {
    int geneCount;
    Sheep studiedSheep("", "");
    vector<Sheep> sheep = readData(geneCount, studiedSheep);

    for (int i = 0; i < sheep.size(); ++i) {
        sheep[i].geneticSimilarity = getGeneticSimilarity(sheep[i].dnr, studiedSheep.dnr, geneCount);
    }

    sheep = sortSheep(sheep);

    outputData(studiedSheep, sheep);

    return 0;
}

File: 2015m./uzd2./sprendimai/uzd2.py
class Sheep:
    geneticSimilarity = 0
    def __init__(self, name, dnr):
        self.name = name
        self.dnr = dnr

def readData():
    data = open("U2.txt", "r")
    sheepCount, geneCount = next(data).split(" ")
    sheepCount = int(sheepCount)
    geneCount = int(geneCount)
    studiedSheepIndex = int(next(data)) - 1
    studiedSheep = None

    sheep = []

    for i in range(sheepCount):
        sheepData = next(data)
        name = sheepData[:10]
        dnr = sheepData[11:]
        if(i == studiedSheepIndex):
            studiedSheep = Sheep(name, dnr)
            continue
        sheep.append(Sheep(name, dnr))

    data.close()

    return studiedSheep, geneCount, sheep

def outputData(studiedSheep, sheep):
    output = open("U2rez.txt", "w")

    output.write(f"{studiedSheep.name}\n")
    for i in range(len(sheep)):
        output.write(f"{sheep[i].name} {sheep[i].geneticSimilarity}\n")

    output.close()

def sortSheep(sheep):
    size = len(sheep)
    for i in range(size):
        for j in range(i + 1, size):
            if(sheep[i].geneticSimilarity < sheep[j].geneticSimilarity or 
              (sheep[i].geneticSimilarity == sheep[j].geneticSimilarity and
               sheep[i].name > sheep[j].name)):
                sheep[j], sheep[i] = sheep[i], sheep[j]

    return sheep

def getGeneticSimilarity(dnr1, dnr2, geneCount):
    geneticSimilarity = 0
    for i in range(geneCount):
        if(dnr1[i] == dnr2[i]):
            geneticSimilarity += 1
    return geneticSimilarity
    
def main():
    studiedSheep, geneCount, sheep = readData()
    
    for i in range(len(sheep)):
        geneticSimilarity = getGeneticSimilarity(sheep[i].dnr, studiedSheep.dnr, geneCount)
        sheep[i].geneticSimilarity = geneticSimilarity
    
    sheep = sortSheep(sheep)

    outputData(studiedSheep, sheep)

if __name__ == "__main__":
    main()